README.md

```diff
@@ -284,6 +284,43 @@ def change do
 end
 ```
 
+### "Empty" transactions when no changes have been recorded
+
+By default, Carbonite will store the `Carbonite.Transaction` record regardless of whether in the corresponding database transaction any changes were recorded or not. In fact, there is nothing special about the behaviour of the following code:
+
+```elixir
+Ecto.Multi.new()
+|> Carbonite.Multi.insert_transaction(%{meta: %{type: "rabbit_inserted"}})
+|> Ecto.Multi.run(:rabbit, &maybe_insert_rabbit/2)
+|> MyApp.Repo.transaction()
+```
+
+Depending on the behaviour of `maybe_insert_rabbit/2`, the transaction may result in one of these outcomes:
+
+1. The transaction succeeds and the rabbit is inserted. A `Carbonite.Transaction` is recorded alongside a `Carbonite.Change`.
+2. The transaction succeeds but *no rabbit is inserted*. The `Carbonite.Transaction` is recorded but will be empty, that is, not associated to any `Carbonite.Change` records.
+3. The transaction fails and is rolled back. The `Carbonite.Transaction` is not persisted.
+
+Outcome (1) and (3) above are engrained in Carbonite's trigger logic and can not be changed.
+
+The second outcome can be considered "intended behaviour" and there are good reasons for keeping the `Carbonite.Transaction` around. However, if your use-case indicates that you should not store `Carbonite.Transactions` when nothing was changed, you need to:
+
+1. Set the trigger to `INITIALLY DEFERRED` when it is created.
+2. Re-order your application logic ...
+3. ... and only insert the transaction when needed.
+
+```elixir
+# In the migration where the trigger is created
+Carbonite.Migrations.create_trigger(:rabbits, initially: "deferred")
+```
+
+```elixir
+Ecto.Multi.new()
+|> Ecto.Multi.run(:rabbit, &maybe_insert_rabbit/2)
+|> Ecto.Multi.run(:carbonite_transaction, &maybe_insert_carbonite_transaction/2)
+|> MyApp.Repo.transaction()
+```
```

Comment:
Maybe we have to talk about the downsides well, right? Like side-effect transactions that may be rolled back because of the carbonite changes insert not working?

Reply:
Not sure what you mean exactly. Like for example an API call that may now (= with this setup) be executed before the carbonite trigger fails? But isn't that just normal transactional behaviour and people should know about it? Not sure whether this is a clear "downside"... 

Downside for me would be that the `maybe_insert_carbonite_transaction` is likely more complex and I'd probably need a custom one every time I do this trick? :) Idk... TBH this entire patch: I'm not 100% convinced that avoiding the "empty transactions" is worth this trouble. Making it deferrable, sure, but all this additional logic? Meh

Reply:
What I meant is the following. Before, when the trigger failed to insert the changes record (e.g. because of no transaction being inserted) it would rollback everything up to that point. So if your transaction has A,B and C and A and B inserts some changes and C has some side-effects, C would not be executed because the transaction is aready rolling back if the changes can not be inserted. 

So with: 
A - random DB stuff
B - do something that triggers changes, where changes fail to be inserted
C 🐒 side effects

non-deffered:
A
B 💥
(C not executed)

deferred:
A
B
C -> side effect
💥 

A and B are rolled back and the side effect of C is still done.

It is defiantly something people can live with, but I though something along that lines is worth mentioning.

Reply:
:thinking: hmm yeah that's what I thought - wouldn't I know this when I go this extra mile to make the trigger deferred and call `Carbonite.insert_transaction` late in my transaction? Idk :shrug: Any suggestion on what to write?

Reply:
> ⚠️ Be aware that with deferred triggers, errors that happen when writing changes will only appear at the end of the DB transaction and can still lead to the transaction being rolled back.

Reply:
:thinking: 

> errors that happen when writing changes

IMO this is a bit too broad. The trigger function only has one "expected" error, which is the FK violation, i.e. a change-tracked table was touched but no `Carbonite.Transaction` was inserted. I'd prefer if we didn't warn about the general possibility of errors in the trigger, so we don't seem dubious of our own code quality.

Reply:
> Be aware that with deferred triggers, in case you miss to insert a `Carbonite.Transaction`, the constraint violation will only occur at the end of the transaction when potential side-effects have already been performed.

?

But I'm generally unsure about this warning, it feels like a too big "wet floor" sign. So many words, forcing readers to stop and consciously think about it, whereas the likeliness and potential impact of this danger is comparatively low?

Reply:
> so we don't seem so sceptical of our own code quality.

Hmm, but who knows what people add on top. Maybe there is complicated priviledges set on the carbonite tables in SQL, or other triggers that people are building on top that can fail (like a trigger that triggers on inserts to the change table and sends postgres notifications again. It's a library after all.

Reply:
![](https://media.giphy.com/media/KccoHFiEXjWZW/giphy-downsized-large.gif)

Reply:
> but who knows what people add on top

Well sure, but I'd suppose these hypothetic people building stuff on top of the Carbonite tables know what they're doing? Isn't this a bit far fetched to worry the "normal" Carbonite user about? Also, I'd like to put some more emphasis on the "sceptical of our own code quality" -> People can add on top whatever they want, sure, but we wouldn't generally add warnings about potential **unknown** errors to everything, right? A la "Be aware there may be bugs in our or your code."? This kind of warning would make me feel worried about the library a bit. And especially so when looking at an auditing library.

However, you also seem to be really keen on the warning. So how about we don't warn about any unspecific error case, but just try to "wake up" the innocent, i.e. by pointing out the deferred trigger isn't really "normal library use"? Maybe something whimsical like

> Be aware that while this functionality may seem useful, it is not recommended by the Carbonite team. Instead, we recommend leaving the empty transactions in place, optionally cleaning them up at a later point. Make sure you know how deferred triggers work before going down this path.

Reply:
@andreasknoepfle added the warning to the doc of `create_trigger/2` https://github.com/bitcrowd/carbonite/pull/61/commits/4794b404feca6d554d7737f8202279ac8296989b

approve?

Reply:
Guess it's fine to merge